lms/djangoapps/bookmarks/views.py

```diff
@@ -0,0 +1,301 @@
+        optional_fields = optional_fields_param.split(',') if optional_fields_param else []
+        context['fields'] = DEFAULT_FIELDS + [field for field in optional_fields if field in OPTIONAL_FIELDS]
+        return context
+
+    def get_queryset(self):
+        course_id = self.request.QUERY_PARAMS.get('course_id', None)
+
+        bookmarks_queryset = Bookmark.objects.filter(user=self.request.user)
+
+        if not course_id:
+            return bookmarks_queryset.order_by('-created')
+        try:
+            course_key = CourseKey.from_string(course_id)
+        except InvalidKeyError:
+            log.error("Invalid course id '{course_id}'")
+            return []
+
+        return bookmarks_queryset.filter(course_key=course_key).order_by('-created')
+
+    def post(self, request):
+        """
+        POST /api/bookmarks/v1/bookmarks/
+        """
+        if not request.DATA:
+            error_message = _("No data provided")
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        usage_id = request.DATA.get('usage_id', None)
+        if not usage_id:
+            error_message = _('No usage id provided')
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            usage_key = UsageKey.from_string(usage_id)
+        except InvalidKeyError as exception:
+            log.error(exception.message)
+            message = _(u"Invalid usage id")
+            return Response(
+                {
+                    "developer_message": message,
+                    "user_message": message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+        else:
+            # course_key may have an empty run property.
+            usage_key = usage_key.replace(course_key=modulestore().fill_in_run(usage_key.course_key))
+            course_key = usage_key.course_key
+
+        bookmarks_data = {
+            "usage_key": usage_key,
+            "course_key": course_key,
+            "user": request.user
+        }
+
+        try:
+            bookmark = Bookmark.create(bookmarks_data)
+        except ItemNotFoundError as exception:
+            log.error(exception.message)
+            return Response(
+                {
+                    "developer_message": u"Block with usage_id not found.",
+                    "user_message": _(u"Invalid usage id")
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+        return Response(
+            BookmarkSerializer(bookmark, context={"fields": DEFAULT_FIELDS + OPTIONAL_FIELDS}).data,
```

Comment:
Why are you passing the complete set of fields as the context instead of using the fields the user requested (via get_serializer_context)?